lib/algos/list/delimiterMaster.js

```diff
@@ -39,31 +40,52 @@ class DelimiterMaster extends Delimiter {
     filter(obj) {
         let key = obj.key;
         const value = obj.value;
+
+        /* Skip keys not starting with the prefix or not alphabetically
+         * ordered. */
         if ((this.prefix && !key.startsWith(this.prefix))
                 || (typeof this.NextMarker === 'string' &&
                     key <= this.NextMarker)) {
             return FILTER_SKIP;
         }
+
+        /* Skip version keys (<key><versionIdSeparator><version>) if we already
+         * have a master version. */
         const versionIdIndex = key.indexOf(VID_SEP);
         if (versionIdIndex >= 0) {
-            // generally we do not accept a specific version,
-            // we only do when the master version is a PHD version
             key = key.slice(0, versionIdIndex);
-            if (key !== this.prvPHDKey) {
-                return FILTER_ACCEPT; // trick repd to not increase its streak
+            /* - key === this.prvPHDKey is triggered when a master version has
+             *   been accepted for this key,
+             * - key === this.NextMarker is triggered when a listing page ends
+             *   on a accepted obj and the next page starts with a version of
+             *   this object. In that case prvPHDKey is default set to
+             *   undefined (in the constructor) and comparing to NextMarker is
+             *   the only way to know  we should not accept this version.
+             *   This test is not redundant with the one at the beginning of
+             *   this function, we are comparing here the key without the
+             *   version suffix. */
+            if (key === this.prvPHDKey || key === this.NextMarker) {
+                /* master version already filtered */
+                return FILTER_SKIP;
             }
         }
         if (Version.isPHD(value)) {
-            // master version is a PHD version: wait for the next version
-            this.prvPHDKey = key;
-            return FILTER_ACCEPT; // trick repd to not increase its streak
+            /* master version is a PHD version, we want to wait for the next
+             * one:
+             * - Set the prvPHDKey to undefined to not skip the next version,
+             * - return accept to avoid users to skip the next values in range
+             *   (skip scan mechanism in metata backend like Metadata or
+             *   MongoClient). */
+            this.prvPHDKey = undefined;
+            return FILTER_ACCEPT;
         }
         if (Version.isDeleteMarker(value)) {
-            // version is a delete marker, ignore
-            return FILTER_ACCEPT; // trick repd to not increase its streak
+            /* version is a delete marker, ignore and return accept for the
+             * same reason than for PHD version.  */
+            return FILTER_ACCEPT;
```

Comment:
I found sth after my approval.
This point and https://github.com/scality/Arsenal/pull/550/files#diff-d72bb66c5574d68c654584b499fe9b9aR80 return `FILTER_ACCEPT`, then will reset the skip-scan mechanism. Hence, as the skipping of the same common prefix is performed later: https://github.com/scality/Arsenal/pull/550/files#diff-d72bb66c5574d68c654584b499fe9b9aR95, we will list (all) objects that are either `isPHD` (Master) or `isDeleteMarker` (Version) for a comment prefix.

Reply:
To avoid this issue, could we move the block that check `commonPrefixes` (https://github.com/scality/Arsenal/pull/550/files#diff-d72bb66c5574d68c654584b499fe9b9aR89) to be performed before checking `isPHD` (https://github.com/scality/Arsenal/pull/550/files#diff-d72bb66c5574d68c654584b499fe9b9aR72)?

Reply:
Returning FILTER_ACCEPT to reset the skip mechanism here is intentional, we want to avoid the client to skip the next entries. 
We reach these two tests when the first one fails to skip the entry, meaning either the entry key does not have a version part or the key has a version and the key part is different from the previous master accepted.
- If the key does have a version part, it is the master version and it may be a PHD. When it is the case we don't want to call addContents or addCOmmonPrefix with this value and, to be able to find the latest version of this key, we want to avoid the client skipping the next entries in its current key range.
- If the entry key has a version part but the entry has not been skipped with the previous test, it is the same principle, we don't want to add it to result and we want to be sure to have a find the latest version.

This kind of input leads to a buggy result by returning FILTER_SKIP in these cases:
`given the SKIP number to activate the skip scan mechanism is SKIP_NUMBER
a listing page ending with SKIP_NUMBER -1 FILTER_SKIP results
the next listing page starting with an entry either a PHD or a delete marker`
Here, if we return SKIP for the first entry of the second listing, it will trigger the skip scan mechanism and we will miss the version for this key.

Reply:
We investigated with @lamphamsy, and we found out that resetting the skip scan when a PHD/isDeleteMarker version is encountered can make it not working at all, depending on the data inside the bucket.

For example you have 10000 master key in a bucket, but with every 90 master key is a PHD one.
You want to do a listBucket operation, the extension will return FILTER_ACCEPT for each PHD version it encounters, which reset the skip scan. So every 90 master keys the skip scan get resetted, and we end up listing the 10000keys instead of ~100.

I think a good solution would be to have a FILTER_CONTINUE flag, that will go to the next entry without resetting the streakLength counter, and use it when we encounter a PHD/isDeleteMarker version.

Reply:
I add sth to @MathieuCassagne 's example: All the keys have a same prefix, e.g. `prefix1/`. The expected result of a list operation should contain an array of common prefixes that is exactly [`prefix1`]. Thanks to the skip-scan mechanism, we should obtain the expected result after scanning several hundreds (200 in Metadata) of keys.
In the case @MathieuCassagne mentioned above, we have to scan all keys as skip-scan's counter is reset for every a PHD Master key.

Reply:
You re absolutely right, we missed this use case. It should be fixed.

Reply:
my previous comment was addressed to your other comment. about the incorrect listing result.